besskge/dataset.py

```diff
@@ -284,11 +284,71 @@ def build_yago310(cls, root: Path) -> "KGDataset":
             neg_tails=None,
         )
 
+    @classmethod
+    def from_triples(
+        cls,
+        data: NDArray[np.int32],
+        split: Tuple[float, float, float] = (0.7, 0.15, 0.15),
+        entity_dict: Optional[List[str]] = None,
+        relation_dict: Optional[List[str]] = None,
+        type_offsets: Optional[Dict[str, int]] = None,
+        neg_heads: Optional[Dict[str, NDArray[np.int32]]] = None,
+        neg_tails: Optional[Dict[str, NDArray[np.int32]]] = None,
+    ) -> "KGDataset":
+        """
+        Build a dataset from an array of triples.
+
+        :param data:
+            Numpy array of triples [head_id, relation_id, tail_id]. Shape
+            (num_triples, 3).
+        :param split:
+            Optional tuple to set the train/validation/test split.
+        :param entity_dict:
+            Optional entity labels by ID.
+        :param relation_dict:
+            Optional relation labels by ID.
+        :param type_offsets:
+            Offset of entity types
+        :param neg_heads:
+            Optional IDs of negative heads.
+        :param neg_tails:
+            Optional IDs of negative tails.
+
+        :return: Instance of the KGDataset class.
+        """
+        num_triples = data.shape[0]
+        num_train = int(num_triples * split[0])
+        num_valid = int(num_triples * split[1])
+
+        rng = np.random.default_rng()
+        rng.shuffle(data, axis=0)
+
+        triples = dict()
+        triples["train"], triples["valid"], triples["test"] = np.split(
+            data, (num_train, num_train + num_valid), axis=0
+        )
+
+        if not entity_dict:
+            entity_dict = list(map(str, np.unique(data[:, [0, 2]])))
```

Comment:
`entity_dict` and `relation_dict` are Optional in the KGDataset dataclass, I would probably leave them as None in the case where they are not passed as input, instead of creating these fake dictionaries? Although, if you think it's useful to have them, I would refrain from using `np.unique` here, in the corner case where some intermediate IDs (for some reason) never appear in the triple list. I'd just go with `np.arange(n_entity)`, `np.arange(n_relation_type)`, with `n_entity` and `n_relation_type` computed with `max` as you did below.